File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    void uniqueComb(vector<vector<int>> &ans,vector<int> arr,vector<int>& ds,int tar,int i){
        if(tar == 0){
            ans.push_back(ds);
            return;
        }
        if(i == arr.size()) return;
        
        for(int ind=i;ind<arr.size();ind++){
            if(arr[i] > tar) return;
            
            if(i != ind && arr[ind] == arr[ind-1])
                continue;
            
            ds.push_back(arr[ind]);
            uniqueComb(ans,arr,ds,tar-arr[ind],ind+1);
            ds.pop_back();
        }
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<int> ds;
        vector<vector<int>> ans;
        sort(candidates.begin(),candidates.end());
        uniqueComb(ans,candidates,ds,target,0);
        return ans;
    }
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    public void combSum2(int nums[],int target,int ind,List<Integer> list,List<List<Integer>> comb){
        if(target == 0){
            comb.add(new ArrayList<>(list));
            return;
        }

        for(int i=ind;i<nums.length;i++){
            if(target < nums[i]) break;

            if(i != ind && nums[i] == nums[i-1]) continue;

            list.add(nums[i]);
            combSum2(nums,target-nums[i],i+1,list,comb);
            list.remove(list.size()-1);
        }
    }
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> list = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();
        combSum2(candidates,target,0,temp,list);
        return list;
    }
}